tools/oeedger8r/src/Sources.ml

```diff
@@ -762,12 +762,11 @@ let generate_trusted (ec : enclave_content) (ep : Intel.Util.edger8r_params) =
   in
   [
     sprintf "#include \"%s_t.h\"" ec.file_shortnm;
+    "#define OE_NEED_STDC_NAMES";
     "";
     "#include <openenclave/edger8r/enclave.h>";
-    "";
-    "#include <stdlib.h>";
-    "#include <string.h>";
-    "#include <wchar.h>";
+    "#include <openenclave/corelibc/string.h>";
```

Comment:
Why do we want to introduce on dependency on oe specific headers rather than standard headers?

Reply:
Because edger8r-generated code shouldn't depend on libc.

Reply:
> Because edger8r-generated code shouldn't depend on libc.

Can you point me to the design doc that explains why it shouldn't?

Reply:
No I don't believe one exists :smile: I'm happy to document it somewhere though.

Reply:
Note that what I meant (and have since edited) is that the generated code on the enclave side shouldn't depend on libc. On the host side its fine.

I'll detail why in documentation in a separate PR.

Reply:
edger8r generated code is intended primarily to be compiled in by an SDK user. 
There will be far more enclave applications using EDL than the few internal libraries that use EDL.

It should be fine for edger8r generated code to depend on standard headers; however if there is a compelling reason for it not to, we can evaluate it.

An aside, if oe provides headers that resemble standard C headers (e.g string.h, whcar.h) but do different things, then it becomes hard to reason about what edger8r generated code does.

Reply:
I don't really understand the objection.

> edger8r generated code is intended primarily to be compiled in by an SDK user.
There will be far more enclave applications using EDL than the few internal libraries that use EDL.

I am referring strictly to edger8r *generated* glue code. It doesn't matter what code enclave developers write in the ecall/ocall code. The problem is that using libc in generated code forces an enclave to depend on libc even if their application doesn't need it.

> It should be fine for edger8r generated code to depend on standard headers; however if there is a compelling reason for it not to, we can evaluate it.

Let me ask a different question. Is there a compelling reason the generated glue code needs to depend on libc?

> An aside, if oe provides headers that resemble standard C headers (e.g string.h, whcar.h) but do different things, then it becomes hard to reason about what edger8r generated code does.

Well we already do. For example `oe_strlen` is used all over the place in our codebase and is in openenclave/corelibc/string.h. The reality is we've re implemented small parts of libc for use in the enclave core so that not all enclave application need to depend on musl. We could rename all these headers to avoid confusion but its a big API breaking change so we'll need to do it more systematically.

Reply:
@jhand2 I would prefer that edger8r always use the `oe_` prefixed names instead of relying on `#define OE_NEED_STDC_NAMES`. Even though it works in the wrapper today, that mechanism relies on inlining to eliminate the potential symbol collision (i.e. compiler needs to elide the corelibc passthrough implementation of `wcslen()` and only generate `oe_wcslen()`). 

It was primarily designed to enable oecore to use 3rdparty code with std libc dependencies without modification (e.g. dlmalloc) If the intention is that oeedger8r should use corelibc implementations, then let's use them explicitly and reduce the number of things that could go wrong.

Reply:
@jhand2 

oeedger8r generated code has been designed to be as much self-contained as possible. The reason is that edge functions are at the security boundary and we want them to be stable no matter how much rest of OE SDK changes.

oecore is NOT a stable interface; neither is it formally defined nor does it make any guarantees about stablity. Standard C library is well defined and does not change.

An example: Recently we decided to no longer publish safemath/safecrt headers (5b8ebe1). Though this PR changed the public API surface, it had zero impact on oeedger8r since the generated code had been designed to be self-contained and had its own safe math functions.

Yes. It is a good goal that edl/oeedger8r does not depend on standard C.
However it should not come at the expense of making edl/oeedger8r depend on unstable interfaces (oecore).

Currently, edger8r generated code has been designed to be self-contained with a minimal bit of dependency on standard C library. This is the reason the same edger8r generated code works on SGX as well as OPTEE. The generated code depends on
-  C header files
- functions that are part of free standing environment (e.g memcpy)
- string functions (strlen, wcslen etc)
- published oe headers

If the dependency (though tiny) with the C library has to be removed, it is preferable to continue to keep oeedger8r generated code self-contained. Instead of depending on OE core
- provide inline functions `oe_edl_strlen` that the generated code can call
- define types in EDL headers rather than oe core headers
- change edger8r to not  emit `#include <wchar.h>` if the EDL does not use wide characters.

Note: It is impossible to not depend on C library functions. Even for a `free standing` environment where no C library is available, GCC is free to generate memcpy, memset and other functions.

Reply:
Thanks for the background! I'll rework this to ensure the generated code stays self contained.

Reply:
Thanks @jhand2 for incorporating comments.

Reply:
I didn't realize that the compiler would just fill in the implementations for stdc functions. Thanks for the heads up @anakrish. I've updated the PR to be a lot more scoped down and just address shrinking the need for the C wrappers.